TF-M: directory structure update proposal to import twincpu branch to mbed-os

urutva opened the pull request:
### Description
In mbed-os, we currently support v8-M targets using TF-M Secure Partition Manager (SPM) and mbed-os services and twin v7-M targets using MBED SPM and mbed-os services. This combination adds significant maintenance overhead. Therefore we are planning use both PSA SPM and services from TF-M to support v8-M and twin v7-M targets.

This PR proposes a new directory structure which is required to import TF-M twin cpu implementation into mbed-os. 

At the time of creating this PR, TF-M master only supports v8-M targets and twin-cpu targets are supported on a feature branch.  Therefore our plan is,

1. Continue supporting v8-M targets on mbed-os master (TF-M SPM and mbed-os services)
2. Continue supporting twin v7-M targets on mbed-os master (MBED SPM and mbed-os services)
3. Continue supporting single v7-M targets on mbed-os master (emulation and mbed-os services)
4. On feature-twincpu branch, add support for twincpu (TF-M SPM and TF-M PSA services)
5. On feature-twincpu branch, maintain support for single v7-M targets on mbed-os master (emulation and mbed-os services)
6. Merge feature-twincpu to master once it is stable and TF-M feature-twincpu branch gets merged with TF-M master branch

### Pull request type
    [ ] Fix
    [x] Refactor
    [ ] Target update
    [ ] Functionality change
    [ ] Docs update
    [ ] Test update
    [x] Breaking change

### Reviewers
@Patater @jainvikas8 

### Release Notes

1. Switch from MBED Secure Partition Manager to TF-M Secure Partition Manager to support twincpu v7-M targets
2. Switch from Mbed secure services to TF-M secure services to support twincpu v7-M targets

#### Impact Analysis
Non-secure side application:
There is no impact on non-secure side application as the PSA APIs will remain the same.
Secure side application:
Breaking change if someone wants to add their own secure service, but we don't provide any migration for them. They need to use TF-M (https://git.trustedfirmware.org/trusted-firmware-m.git/tree/) instead.

urutva committed:
Remove MBED SPM implementation and test cases

Signed-off-by: Devaraj Ranganna <devaraj.ranganna@arm.com>

urutva committed:
Remove imported TF-M SPM related files

These changes are the first step towards switching to TF-M SPM and
services which will be running on the secure side and mbed-os
will be running on the non-secure side.

Signed-off-by: Devaraj Ranganna <devaraj.ranganna@arm.com>

urutva committed:
Refactor Mbed PSA service code base

The Mbed OS PSA services is moved to
features/FEATURE_PSA/TARGET_MBED_PSA_SRV.

Removed PSA_SRV_IMPL, PSA_SRV_EMUL and NSPE.

Added new feature `PSA` to support PSA in Mbed OS.

Created following generic PSA targets:

* `PSA_Target` (Root level PSA target)
* `PSA_V7_M_NSPE` (Single v7-M NSPE generic target)
* `PSA_V7_M_SPE` (Single v7-M SPE generic target)
* `PSA_DUAL_V7_M_NSPE` (Dual v7-M NSPE generic target)
* `PSA_DUAL_V7_M_SPE` (Dual v7-M SPE generic target)
* `PSA_V8_M_NSPE` (v8-M NSPE generic target)
* `PSA_V8_M_SPE` (v8-M SPE generic target)

Added document features/FEATURE_PSA/supporting_psa_in_mbed-os.md which
describes PSA support in Mbed OS.

Signed-off-by: Devaraj Ranganna <devaraj.ranganna@arm.com>

urutva committed:
Update .astyleignore to match new directory structure

Signed-off-by: Devaraj Ranganna <devaraj.ranganna@arm.com>

urutva committed:
Cleanup tools folder and update travis file

1. Remove psa related hooks from the build system.
2. Remove PSA related scripts from tools/test folder
3. Remove psa-autogen job from travis which was running
generate_partition_code.py and is not needed anymore
4. Install python modules needed for events and littlefs tests

Signed-off-by: Devaraj Ranganna <devaraj.ranganna@arm.com>

Patater commented on platform/mbed_retarget.cpp:
```diff
@@ -913,7 +913,7 @@ extern "C" long PREFIX(_flen)(FILEHANDLE fh)
 }
 
 // Do not compile this code for TFM secure target
-#if !defined(COMPONENT_SPE) || !defined(TARGET_TFM)
```
Why remove COMPONENT_SPE here?

urutva replied:
@Patater We do not have `COMPONENT_SPE` anymore or `FEATURE_SPE` as we'll not build secure side using mbed-cli. But in this case we need to identify non-secure and the current #if i snot doing that. I'll change the code to `#if defined(FEATURE_NSPE)` which is only defined for NSPE.

Patater commented on tools/test/psa/__init__.py:
```diff
@@ -1,17 +0,0 @@
-"""
```
Commit "Remvoe PSA related scripts from tools/test folder" should say "Remove PSA related scripts from tools/test folder"

urutva replied:
I'll squash the commits and update the comment.

Patater commented on .travis.yml:
```diff
@@ -196,6 +196,17 @@ matrix:
+        - python -m pip install --upgrade pip==18.1
+        - python -m pip install --upgrade setuptools==40.4.3
+        - pip install -r requirements.txt
+        - pip list --verbose
```
How was the events test working previously without these changes?

urutva replied:
They were included as part of psa-autogen test. The events test failed when I removed them.

jainvikas8 commented on .travis.yml:
```diff
@@ -1,4 +1,4 @@
-# Copyright (c) 2013-2018 Arm Limited. All rights reserved.
```
Please fix up the commit message.

urutva replied:
@jainvikas8 what's missing from the commit message?

jainvikas8 reviewed (commented):
Heading less than 50 characters, please.

Patater commented on .travis.yml:
```diff
@@ -194,9 +194,8 @@ matrix:
-      stage: "Extended"
-      name: "psa autogen"
-      env: NAME=psa-autogen
+      name: "events"
```
The events test previously ran under `stage: "Extended"`. This should probably come back.

urutva replied:
Fixed.

Patater commented on tools/config/__init__.py:
```diff
@@ -69,7 +69,7 @@
 
 ALLOWED_FEATURES = [
-    "BOOTLOADER", "BLE", "LWIP", "STORAGE", "NANOSTACK", "CRYPTOCELL310",
+    "BOOTLOADER", "BLE", "LWIP", "STORAGE", "NANOSTACK", "CRYPTOCELL310", "PSA", "TFM", "TFM_TWINCPU", "TFM_V8M", "MBED_PSA_SRV", "PSA_SRV_IPC", "PSA_SRV_IMPL", "PSA_SRV_EMUL", "NSPE",
```
Why isn't Mbed TLS listed as an allowed feature here? It's present in `features/mbedtls`. If this list isn't for enumerating all possible features, what's this list for?

urutva replied:
@Patater I'm not sure why `features/mbedtls` is not listed here. @mark-edgeworth any thoughts.
if `features_add` config option is mentioned in `targets.json` then https://github.com/ARMmbed/mbed-os/blob/master/tools/config/__init__.py#L1347 is invoked which checks the added feature against `ALLOWED_FEATURES`.

mark-edgeworth replied:
Sorry, no. This is not something that we'd be involved with.

0xc0170 replied:
If we change this, this doc should be updated : https://github.com/ARMmbed/mbed-os-5-docs/blob/46e78f5b7b8c38c383820af3adc9a85d758c0c75/docs/reference/configuration/mbed_targets.md#features-features_add-and-features_remove

I dont think this change makes sense, so many features ? They do not look to me like a complete feature (for instance if I use TFM why do I need to add also PSA and the rest) ?

0xc0170 replied:
This list looks outdated though, lwip and storage and nanostack are not features anymore. same as psa/tfm should not be?

Patater replied:
> They do not look to me like a complete feature

So should these be components? :p

There are so many of them so that we can control the globbing build system well enough. We should have some documentation about how these filters all work (and would need to update it as well, along with all other Mbed OS + PSA Docs), although I can't find the component explaining docs right now. @devran01 Do we have a link we could share with @0xc0170 ?

urutva replied:
Thanks for the summary @Patater . @0xc0170 Unfortunately, these are not complete features but something that we need to support single v7-M, dual v7-M and v8-M targets. 
The document describing the component structure is https://github.com/ARMmbed/mbed-os-5-docs/blob/development/docs/porting/psa/spm.md#defining-a-psa-compliant-target-in-targetsjson.
Some more details https://github.com/ARMmbed/mbed-os/pull/11640#issuecomment-540997520.

I think it's better to update `mbed-os-docs` once we merge this feature branch onto master. But before that we need to agree on adding PSA as a feature and accept having incomplete features combined together or revert back to components as before.

0xc0170 replied:
not components. But as there are so many. Why are these not coming from targets itself (part of config) ? There are like 8 of them added here. How can I combine them (is it even possible to just use one or two?) ?

Why would I care about any of these to be added as "feature" ?:
""PSA", "TFM", "TFM_TWINCPU", "TFM_V8M", "MBED_PSA_SRV", "PSA_SRV_IPC", "PSA_SRV_IMPL", "PSA_SRV_EMUL", "NSPE""

most of them have TFM or PSA prefix, Why if I want PSA these are not coming in as well? Or TFM? I do not understand how they are related so take my comment as a outsider :)

This table illustrates it for me: https://github.com/ARMmbed/mbed-os-5-docs/blob/development/docs/porting/psa/spm.md#defining-a-psa-compliant-target-in-targetsjson

Table shows label vs core. Why I cant override atarget to use one of these if I want? Using config (target_overrides) or have it predefined in target config ?

Patater commented on tools/psa/__init__.py:
```diff
@@ -20,9 +20,6 @@
-
-
 
 def find_secure_image(notify, resources, ns_image_path,
```
Who uses `find_secure_image()` still? I am curious if we need any files in `tools/psa` at all.

urutva replied:
@Patater Currently, `find_secure_image()` is not needed. Having said that, once we add tf-m binaries to targets then we need this method to find the secure image and merge it with non-secure image. It's better to retain it.
The python scripts under `tools/psa/tfm` are needed to sign the combined binary, but they are not up to date. I can delete the folder now and bring those files back when we import the NS APIs from tf-m using importer script or retain the folder. Any recommendation?
I'll delete `partition_description_schema.json` and `README.md`.

urutva replied:
@Patater any recommendation?

Patater replied:
@jainvikas8 Mentioned that this would be needed for any v8-M targets, which we don't care to support from this branch.

Please correct me if I'm wrong, but I don't believe we'll need it to build the combined image if we place the binary in the proper location in the Mbed OS tree. No special "find" step should be needed as the globbing build system will pull in everything it can. I'm thinking the SPE binaries would go in a similar location (similar filter folders) to the linker script for each target.

urutva replied:
@Patater Got it. Removed `tools/psa` folder.

jainvikas8 commented on features/FEATURE_PSA/supporting_psa_in_mbed-os.md:
````diff
@@ -0,0 +1,189 @@
+```
+
+Please pay attention to config options `extra_labels_add` and `device_has_add`. If needed then PSA target defintion **MUST** use [extra_labels/device_has]`_add` or [extra_labels/device_has]`_remnove` to add/remove either extra_labels or target capabilities. Also, use [feature]`_add` or [feature]`_remove` to add/remove a feature.
+Check [extra_lables](https://github.com/ARMmbed/mbed-os-5-docs/blob/development/docs/reference/configuration/mbed_targets.md#extra_labels-extra_labels_add-and-extra_labels_remove), [device_has](https://github.com/ARMmbed/mbed-os-5-docs/blob/development/docs/reference/configuration/mbed_targets.md#device_has) and [features](https://github.com/ARMmbed/mbed-os-5-docs/blob/development/docs/reference/ ...
````
**[extra_labels/device_has]`_remove`**

urutva replied:
Fixed

urutva committed:
Remove v8-M targets

Removed MUSCA A1 and LPC55S69 targets as they are not supported on
feature branch. The targets will be re-introduced before merging feature
branch to master.

Signed-off-by: Devaraj Ranganna <devaraj.ranganna@arm.com>

Patater commented on features/FEATURE_PSA/supporting_psa_in_mbed-os.md:
```diff
@@ -0,0 +1,202 @@
+# Supporting PSA in Mbed OS
+
+This document is still work in progress. It describes the process of adding new PSA targets to Mbed OS. It also outlines the new directory structure created to support PSA on single v7-M, dual v7-M and single v8-M.
```
```diff
- This document is still work in progress.
+ This document is still a work in progress.
```

urutva replied:
Fixed

Patater commented on features/FEATURE_PSA/supporting_psa_in_mbed-os.md:
````diff
@@ -0,0 +1,202 @@
+```json
+"inherits": ["Target"],
+```
+to PSA target definition.
````
```diff
- to PSA target definition.
+ to the new PSA target definition.
```

urutva replied:
Fixed

Patater commented on features/FEATURE_PSA/supporting_psa_in_mbed-os.md:
```diff
@@ -0,0 +1,202 @@
+to PSA target definition.
+
+### Adding single v7-M targets
+The PSA on single v7-M is supported with the help of Mbed OS PSA services. The following is taken from `targets/targets.json` and shows PSA enabled target `K64F`.
```
```diff
- shows PSA enabled target `K64F`.
+ shows a PSA enabled target, `K64F`.
```

urutva replied:
Fixed

Patater commented on features/FEATURE_PSA/supporting_psa_in_mbed-os.md:
````diff
@@ -0,0 +1,202 @@
+    }
+```
+
+Please pay attention to config options `extra_labels_add` and `device_has_add`. If needed then PSA target definition **MUST** use [extra_labels/device_has]`_add` or [extra_labels/device_has]`_remove` to add/remove either extra_labels or target capabilities. Also, use [feature]`_add` or [feature]`_remove` to add/remove a feature.
````
This means we must not use "extra_labels" or "device_has" for PSA targets, correct? If so, I think it's worth saying so explicitly here.

```diff
- If needed then PSA target definition **MUST** use [extra_labels/device_has]`_add` or [extra_labels/device_has]`_remove`
+ If needed then a PSA target definition **MUST** use [extra_labels/device_has]`_add` or [extra_labels/device_has]`_remove` (not `extra_labels` or `device_has`)
```

urutva replied:
Agree. Fixed.

Patater commented on features/FEATURE_PSA/supporting_psa_in_mbed-os.md:
```diff
@@ -0,0 +1,202 @@
+### Adding v8-M targets
+
+##  Enabling PSA at application level
+Having an entropy source is crucial for Mbed TLS and Mbed-Crypto. The [document](https://os.mbed.com/docs/mbed-os/v5.14/porting/entropy-sources.html) talks about entropy and how to add an entropy source. Sometimes a target might not have a True Random Number Generator (TRNG) and in that case the target will be configured as a non-PSA target in `targets/targets.json`. In that scenario, if an appli ...
```
```diff
- Mbed-Crypto
+ Mbed Crypto
```

Patater replied:
```diff
- if an application wants to use that target as a PAS target
+ if an application wants to use that target as a PSA target
```

Patater replied:
```diff
- can be used to make a target PSA target.
+ can be used to make a target into a PSA target.
```

or

```diff
- can be used to make a target PSA target.
+ can be used to enable PSA for a target.
```

urutva replied:
Fixed.

Patater commented on features/FEATURE_PSA/supporting_psa_in_mbed-os.md:
````diff
@@ -0,0 +1,202 @@
+```
+to PSA target definition.
+
+### Adding single v7-M targets
````
The method described in this document doesn't seem to match what's present in this PR's targets.json. Some targets use `"features_add": ["PSA"],` instead of inheriting from a PSA target. Could you extend this document to explain when we should prefer `features_add` to get PSA?

urutva replied:
@Patater Good catch. The method described in the document should be used while creating new PSA targets. Because, just adding `"features_add": ["PSA"]` won't bring in PSA services for v7-M (`"extra_labels": ["MBED_PSA_SRV"]`). I've updated targets.json.

bulislaw commented on targets/targets.json:
```diff
@@ -2092,113 +2042,6 @@
         "device_has_remove": ["QSPI"],
         "release_versions": ["2", "5"]
     },
-    "LPC55S69": {
```
Why is that removed?

Patater replied:
This would obviously not be acceptable on `master`. This changeset is targeting a feature branch which aims only to work with PSoC 6. Removing LPC55S69 from targets.json was the most straightforward way to disable tests for targets we are deliberately breaking (and not yet fixing). When we take the branch back into `master`, we'll need to do the work to ensure this target works.

urutva committed:
Disable mbed-os-example-atecc608a compilation

The example `mbed-os-example-atecc608a` needs the secure element driver
which is removed from feature branch. The secure element support should
be added to TF-M and when we import TF-M to mbed-os, secure element
will be supported in mbed-os. Because of this compilation of the
example `mbed-os-example-atecc608a` is disabled.

Note: DO NOT merge this to `master` without secure element support in
mbed-os.

Signed-off-by: Devaraj Ranganna <devaraj.ranganna@arm.com>

urutva committed:
Improve multiple inheritance support

Current algorithm doesn't support multiple inheritance when inheritance
depth is more than 1 on two parents. It'll not parse the "_add" or
"_remove" attributes on a second parent which is at the same level
as that of a parent that defines the attribute.

This change ensures that second parent which is at the same level
as that of a parent that defines the attribute is parsed for "_add" or
"_remove" attributes. However, this change will parse the parent which
defines the attributes twice, once to evaluate the attribute and then to
evaluate the "_add" or "_remove" attribute. But, no target is allowed to
define an attribute and also "_add" or "_remove" for the same attribute.

Signed-off-by: Devaraj Ranganna <devaraj.ranganna@arm.com>

Patater commented on targets/targets.json:
```diff
@@ -2042,113 +2042,6 @@
         "device_has_remove": ["QSPI"],
         "release_versions": ["2", "5"]
     },
-    "LPC55S69": {
```
Commit "Remove v8-M and dual v7-M targets" seems only to remove v8-M targets. Consider changing commit message.

urutva replied:
Yeah, rebasing on top master had removed dual v7-M targets. Updated the commit message.

Patater commented on tools/test/psa/test_find_secure_image.py:
```diff
@@ -1,57 +0,0 @@
-# Copyright (c) 2019 ARM Limited
```
If you are keeping the `find_secure_image()` function, shouldn't we also keep the test?

urutva replied:
Yes, missed that. I'll bring back the test.

Patater commented on targets/targets.json:
```diff
@@ -9183,31 +9132,6 @@
         },
         "program_cycle_s": 10
     },
-    "CY8CKIT_062_WIFI_BT": {
```
Why remove this target in commit "Refactor Mbed PSA service code base "?

urutva replied:
The target shouldn't have been removed. Fixed.
